astropy/utils/data.py

```diff
@@ -1327,43 +1402,300 @@ def _get_download_cache_locs():
+    check_hashes : bool, optional
+        Whether to compute the hashes of the contents of all files in the
+        cache and compare them to the names. This can take some time if
+        the cache contains large files.
+
+    Returns
+    -------
+    strays : set of strings
+        This is the set of files in the cache directory that do not correspond
+        to known URLs. This may include some files associated with the cache
+        index.
+
+    Raises
+    ------
+    CacheDamaged
+        To indicate a problem with the cache contents; the exception contains
+        a .bad_url or a .bad_file attribute to allow the user to use
+        `clear_download_cache` to remove the offending item.
+    OSError, RuntimeError
+        To indicate some problem with the cache structure. This may need a full
+        `clear_download_cache` to resolve, or may indicate some kind of
+        misconfiguration.
+    """
+    # We're only reading but without the lock goodness knows what
+    # inconsistencies might be detected
+    with _cache(write=True) as (dldir, url2hash):
+        leftover_files = set(os.path.join(dldir, k)
+                             for k in os.listdir(dldir))
+        try:
+            leftover_files.remove(os.path.join(dldir, "urlmap"))
+        except KeyError:
+            pass
+        try:
+            leftover_files.remove(os.path.join(dldir, "lock"))
+        except KeyError:
+            raise RuntimeError("Lock file missing!?")
+        for u, h in url2hash.items():
+            if not os.path.exists(h):
+                msg = "URL '{}' points to nonexistent file '{}'".format(
+                        u, h)
+                raise CacheDamaged(msg, bad_url=u)
+            try:
+                leftover_files.remove(h)
+            except KeyError:
+                # Huh. Two different URLs retrieved identical files.
+                pass
+            d, hexdigest = os.path.split(h)
+            if dldir != d:
+                msg = ("Expected downloaded files to be in '{}' but "
+                       "URL '{}' points to file '{}'"
+                       .format(dldir, u, h))
+                raise CacheDamaged(msg, bad_url=u)
+            if check_hashes:
+                hexdigest_file = compute_hash(h)
+                if hexdigest_file != hexdigest:
+                    msg = ("File corresponding to {} has contents that "
+                           "do not match the hash value: '{}' "
+                           .format(u, hexdigest))
+                    raise CacheDamaged(msg, bad_file=h)
+        for h in leftover_files:
+            h_base = os.path.basename(h)
+            if len(h_base) >= 32 and re.match("[0-9a-f]+", h_base):
+                raise CacheDamaged("Apparently abandoned hash file {}"
+                                   .format(h), bad_file=h)
+    return leftover_files
+
+
+def import_to_cache(url_key, filename,
+                    hexdigest=None,
+                    remove_original=False):
```

Comment:
I think the default should be `remove_original=True`, right? Otherwise, it would take twice as much disk space now than before. Or should this be a configuration setting that user can control; why not use `conf.delete_temporary_downloads_at_exit`?

Reply:
For normal downloading, it's definitely used with remove_original=True. If the user has control of the URL that is being requested, indeed they can just supply a file:// URL to download_cache (with cache=False) or probably use some other method to avoid the cache entirely. But the import_to_cache is meant for situations where the original isn't readily available but the user has somehow obtained the file and stored it locally. So I'm hesitant to delete a file the user hands us without some positive indication.